doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,882 @@
+* Currently, if a user launches the terminal as an administrator, then _all_ of
+  their tabs run elevated. We'll probably want to provide a similar
+  configuration as well. Maybe if a window process is initially launched as
+  elevated (rather than inheriting from an existing session), then it could
+  either:
+  - default all future connections it creates to elevated connections
+  - assume the _first_ tab (or all the terminals created by the startup
+    commandline) should be elevated, but then subsequent connections would
+    default to unelevated processes.
+
+It's a long-standing platform bug that you cannot use drag-and-drop in elevated
+scenarios. This unfortunately means that elevated window processes will _not_ be
+able to tear tabs out into their own windows. Even tabs that only contain
+unelevated content processes in them will not be able to be torn out, because
+the drag-and-drop service is fundamentally incapable of connecting to elevated
+windows.
+
+We should probably have a discussion about what happens when the last elevated
+content process in an elevated window process is closed. If an elevated window
+process doesn't have any remaining elevated content processes, then it should be
+able to revert to an unelevated window process. Should we do this? There's
+likely some visual flickering that will happen as we re-create the new window
+process, so maybe it would be unappealing to users. However, there's also the
+negative side effect that come from elevated windows (the aforementioned lack of
+drag/drop), so maybe users will want to return to the more permissive unelevated
+window process.
+
+This is one section where unfortunately I don't have a working prototype yet.
+From my research, an elevated process cannot simply `create_instance` a class
+that's hosted by an unelevated process. However, after proposing this idea to
+the broad C++/WinRT discussion alias, no one on that thread immediately shot
+down the idea as being impossible or terrible from a security perspective, so I
+believe that it _will_ be possible. We still need to do some research on the
+actual mechanisms for some token impersonation.
+
+### Monarch and Servant Processes
+
+With the current design, it's easy to connect many content processes to a window
+process, and move those content processes easily between windows. However, we
+want to make sure that it's also possible to communicate between these
+processes. What if we want to have only a single WT instance, so that whenever
+Wt is run, it creates a tab in the existing window? What if you want to run a
+commandline in a given WT window?
+
+In addition to the concept of Window and Content Processes, we'll also be
+introducing another type of categorization for window processes. These are
+"Monarch" and "Servant" processes. This will allow for the coordination across
+the various windows by the Monarch.
+
+There will only ever be one monarch process at a given time, and every other WT

```

Comment:
What happens if you only have one window process, that is the monarch, and it throws an exception for whatever reason?

Will the content processes also attempt to be servants here? Or perhaps is there another mechanism by which the content processes can realize all the windows are gone and start a new window process who becomes the instant-monarch for all the still-living and remaining content processes?

Reply:
See, that's tricky. I suppose we _could_ have a content process try to auto-create a new window for itself if it detects that it's window is gone, but it didn't receive a clean close. Like, the window process would _need_ to call `Content::WindowIsClosing()` to indicate that it's a clean close manually. I'm not sure from and end user perspective if all the panes in a window exploding out into their own windows is totally a great idea, but it's worth noting as a possibility.

Reply:
(I've added a discussion in "Potential Issues" for this. I'm not sure that we need to have an answer for this scenario now, but it's something that we could try.